google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/ManagedCloudSdkService.java

```diff
@@ -145,71 +124,57 @@ void invokeOnApplicationUIThread(Runnable runnable) {
     ApplicationManager.getApplication().invokeLater(runnable);
   }
 
-  /**
-   * Checks for Managed Cloud SDK status and creates/installs it if necessary.
-   *
-   * @return Path of the installed managed SDK, null if error occurred while installing.
-   * @throws InterruptedException if Managed Cloud SDK has been interrupted by {@link
-   *     #cancelInstall()}
-   */
-  // TODO(ivanporty) simplify exception handling once all their use cases are clear.
-  private Path installManagedSdk() throws InterruptedException {
+  /** Creates managed SDK and checks for fatal errors. */
+  @VisibleForTesting
+  void initManagedSdk() {
     if (managedCloudSdk == null) {
       try {
         managedCloudSdk = createManagedSdk();
       } catch (UnsupportedOsException ex) {
         logger.error("Unsupported OS for Managed Cloud SDK", ex);
         updateStatus(SdkStatus.NOT_AVAILABLE);
-        return null;
       }
     }
+  }
 
-    updateStatus(SdkStatus.INSTALLING);
-
-    Path installedManagedSdkPath;
-    try {
-      installedManagedSdkPath = installSdk();
-    } catch (InterruptedException iex) {
-      updateStatus(SdkStatus.NOT_AVAILABLE);
-      throw iex;
-    } catch (UnsupportedOsException ex) {
-      logger.error("Unsupported OS for Managed Cloud SDK", ex);
-      updateStatus(SdkStatus.NOT_AVAILABLE);
-      return null;
-    } catch (Exception ex) {
-      logger.error("Error while installing managed Cloud SDK", ex);
-      updateStatus(SdkStatus.NOT_AVAILABLE);
-      return null;
+  /**
+   * Runs managed SDK install/update code on background job and handles success/errors.
+   *
+   * @param jobType Install/update
+   * @param managedSdkTask Task to execute.
+   * @return True of task started, false if Managed SDK is not supported at all.
+   */
+  private boolean executeManagedSdkJob(ManagedSdkJobType jobType, Callable<Path> managedSdkTask) {
+    if (managedCloudSdk == null) {
+      return false;
     }
 
-    try {
-      installAppEngineJavaComponent();
-    } catch (InterruptedException iex) {
-      updateStatus(SdkStatus.NOT_AVAILABLE);
-      throw iex;
-    } catch (Exception ex) {
-      logger.error("Error while installing managed Cloud SDK App Engine Java component", ex);
-      updateStatus(SdkStatus.NOT_AVAILABLE);
-      return null;
+    if (managedSdkBackgroundJob == null || managedSdkBackgroundJob.isDone()) {
+      updateStatus(SdkStatus.INSTALLING);
```

Comment:
there doesn't need to be an SdkStatus#UPDATING?

Reply:
So far I haven't thought of a specific use case for install vs. update for SDK service external users. `INSTALLING` says it's busy until `READY`. The only difference might be for messages we're showing to the user - how about we leave it like this for now, and when we implement UI checking for status, decide if we need additional status?

Reply:
sounds good